src/groups/mqb/mqbauthn/mqbauthn_authenticationcontroller.cpp

```diff
@@ -0,0 +1,483 @@
+            errorDescription << "Duplicate authenticator mechanism '"
+                             << normMech << "' for plugin '" << pluginName
+                             << "'. Each mechanism can only "
+                             << "have one active authenticator";
+            return rc_DUPLICATE_MECHANISM;  // RETURN
+        }
+
+        // Start the authenticator
+        if (int status = authenticator->start(errorStream)) {
+            BMQTSK_ALARMLOG_ALARM("#AUTHENTICATION")
+                << "Failed to start Authenticator '" << authenticator->name()
+                << "' [rc: " << status << ", error: '" << errorStream.str()
+                << "']" << BMQTSK_ALARMLOG_END;
+            errorStream.reset();
+            return status * 10 + rc_START_AUTHENTICATOR;  // RETURN
+        }
+
+        BALL_LOG_INFO << "Started authenticator '" << authenticator->name()
+                      << "' with mechanism '" << normMech << "'";
+
+        // Add to collection
+        d_authenticators.emplace(normMech,
+                                 bslmf::MovableRefUtil::move(authenticator));
+    }
+
+    return rc_SUCCESS;
+}
+
+int AuthenticationController::createDefaultAnonAuthenticator()
+{
+    // Create an anonymous pass authenticator
+    AnonAuthenticatorPluginFactory anonFactory;
+    AuthenticatorMp authenticator = anonFactory.create(d_allocator_p);
+
+    bmqu::MemOutStream errorStream(d_allocator_p);
+
+    // Start the anonymous pass authenticator
+    if (int status = authenticator->start(errorStream)) {
+        BMQTSK_ALARMLOG_ALARM("#AUTHENTICATION")
+            << "Failed to start Authenticator '" << authenticator->name()
+            << "' [rc: " << status << ", error: '" << errorStream.str() << "']"
+            << BMQTSK_ALARMLOG_END;
+        errorStream.reset();
+        return status;  // RETURN
+    }
+
+    // Add the authenticator into the collection
+    const bsl::string normMech = normalizeMechanism(authenticator->mechanism(),
+                                                    d_allocator_p);
+    d_authenticators.emplace(normMech,
+                             bslmf::MovableRefUtil::move(authenticator));
+
+    return 0;
+}
+
+int AuthenticationController::validateAnonymousCredential(
+    bsl::ostream&                      errorDescription,
+    const mqbcfg::AuthenticatorConfig& authenticatorConfig)
+{
+    enum RcEnum {
+        rc_SUCCESS                        = 0,
+        rc_NO_MATCHING_AUTHENTICATOR      = -1,
+        rc_FAIL_CREATE_ANON_AUTHENTICATOR = -2
+    };
+
+    const bdlb::NullableValue<mqbcfg::AnonymousCredential>& anonCredential =
+        authenticatorConfig.anonymousCredential();
+
+    if (anonCredential.isNull()) {
+        bsl::string anonMechanism = bsl::string(AnonAuthenticator::k_MECHANISM,
+                                                d_allocator_p);
+
+        if (d_authenticators.find(anonMechanism) != d_authenticators.cend()) {
```

Comment:
```suggestion
        if (d_authenticators.find(AnonAuthenticator::k_MECHANISM) != d_authenticators.cend()) {
```
Possible to avoid constructing tmp string

Reply:
It would cause a compiler error since `bsl::unordered_map<bsl::string, AuthenticatorMp>` doesn't accept `bsl::string_view` as the parameter to `find`.